source/geometries/JaszczakPhantom.cc

```diff
@@ -0,0 +1,200 @@
+
+  this->SetLogicalVolume(cylinder_logic);
+
+
+  auto water_name = "WATER_BCKG";
+  auto water_solid =
+    new G4Tubs(water_name, 0, cylinder_inner_diam_/2, cylinder_height_/2., 0, twopi);
+
+  G4Material* water = G4NistManager::Instance()->FindOrBuildMaterial("G4_WATER");
+  auto water_logic =
+    new G4LogicalVolume(water_solid, water, water_name);
+  G4VPhysicalVolume* water_phys =
+    new G4PVPlacement(0, G4ThreeVector(0., 0., 0.), water_logic, water_name, cylinder_logic,
+                      false, 0, true);
+
+  cyl_gen_ = new CylinderPointSampler2020(water_phys);
+
+  // Spheres
+  std::vector<G4double> sphere_radii =
+    {sphere1_d_/2, sphere2_d_/2, sphere3_d_/2, sphere4_d_/2, sphere5_d_/2, sphere6_d_/2};
+  auto radius_pos = cylinder_inner_diam_/4.;
+  auto z_pos      = - cylinder_height_/2. + sphere_height_;
+
+  for (unsigned long i=0; i<sphere_radii.size(); i++) {
+    BuildSpheres(i, sphere_radii[i], radius_pos, z_pos, water_logic, water);
+  }
+
+  // Rods
+  std::vector<G4double> rod_radii =
+    {rod1_d_/2, rod2_d_/2, rod3_d_/2, rod4_d_/2, rod5_d_/2, rod6_d_/2};
+  z_pos = - cylinder_height_/2. + rod_height_/2;
+  for (unsigned long i=0; i<rod_radii.size(); i++) {
+    BuildRods(i, rod_radii[i], z_pos, water_logic, water);
+  }
+
+  // Relative actvities
+  auto max_activity = std::max(sphere_activity_, std::max(bckg_activity_, rod_activity_));
+  bckg_activity_   /= max_activity;
+  sphere_activity_ /= max_activity;
+  rod_activity_    /= max_activity;
+
+  G4cout << "*** Relative activities (background, spheres, rods) ***" << G4endl;
+  G4cout << bckg_activity_ << ", " << sphere_activity_  << ", " << rod_activity_ << G4endl;
+
+}
+
+
+void JaszczakPhantom::BuildSpheres(unsigned long n, G4double r, G4double r_pos, G4double z_pos,
+                                   G4LogicalVolume* mother_logic, G4Material* mat) const
+{
+  auto sphere_name = "SPHERE" + std::to_string(n);
+  auto sphere_solid = new G4Orb(sphere_name, r);
+  auto sphere_logic = new G4LogicalVolume(sphere_solid, mat, sphere_name);
+  auto angle = n * 60 * deg;
+  auto x_pos = r_pos * cos(angle);
+  auto y_pos = r_pos * sin(angle);
+  new G4PVPlacement(0, G4ThreeVector(x_pos, y_pos, z_pos), sphere_logic,
+                    sphere_name, mother_logic, false, 0, true);
+}
+
+
+
+void JaszczakPhantom::BuildRods(unsigned long n, G4double r, G4double z_pos,
+                                G4LogicalVolume* mother_logic, G4Material* mat) const
+  {
+    auto diam = 2 * r;
+
+    // Sector displacement from centre, to accommodate gap between sectors
+    auto gap =  14.4 * mm;
+    auto dx  = gap * cos(pi/6);
+    auto dy  = gap * sin(pi/6);
+    // Displacement of first rod WRT sector corner
+    dx += r * sqrt(3);
+    dy += r;
+    // Basis vectors of rod lattice
+    const auto Ax = 2.0, Ay = 0.0;
+    const auto Bx = 1.0, By = sqrt(3);
+    auto a = 0;
+    for (bool did_b=true ; did_b; a+=1) {
+      did_b = false;
```

Comment:
It is not clear to me what `did_b ` is and how the two loops work.

Reply:
Same as above.

Reply:
But you don't understand it either? I printed `did_b` and everytime it is `true` until line 162, but there is no condition for it, while `a` is increasing until it reaches line 167 and breaks. I still don't see the point of `did_b`.

Reply:
This is how the `for (bool did_b=true; did_b; a+=1)` loop works: the first element inside the parenthesis is the starting value of the variable, the second element is the condition that keeps the loop going (meaning, when `did_b` is `false`, the loop stops, and the third element is executed in each iteration of the loop (meaning, `a` increases its value by 1). 

As far as I understand it, `did_b` becomes `false` at the beginning of each iteration of the outer loop. Then the inner loop starts, and at the end of each iteration `did_b` becomes `true` again. During the inner loop, the rods of each line are instantiated, and the loop breaks when it reaches the border of the cylinder or the other condition I added. When it breaks, we go back to the second iteration of the outer loop. If at least one rod has been instantiated, `did_b` is `true`, therefore the loop goes over the next iteration, because it fulfils the condition `did_b`. If not, the loop stops. This last case happens when in the inner loop no rods have been instantiated at all.

Reply:
Now I see, thank you!